### PR TITLE
Normedumbrella

### PR DESCRIPTION
**Pull request by aniemeyer**
This adds functions to compute normed umbrella descriptors of surfaces and for closed surfaces without ears also the ability to reconstruct the umbrella descriptors for the sequence of vertex degrees of a normed umbrella descriptor

### CODE REVIEW

File: gap/PolygonalComplexes/constructors.gi [hunk 1]
```diff
@@ -941,8 +1027,650 @@ end);
...
+        # is valid until row c and position p if face F
+        # were to be inserted there. It assumes it is valid until then
+        isValid := function(ud, c, F, p )
+            local u, r, l, nbs, i, G, neighs, neighsG, j, oncyc, alln;
+
+            u := ud[c]; l := ucode[c][1];
+
+            # if F already occurs in u then we cannot add it again
+            if Pos(u,l,F)  <> false then
+#                Print("face ", F, " already present\n");
```
Comment (MeikeWeiss) [inline gap/PolygonalComplexes/constructors.gi:1455]: This line can be removed.

File: gap/PolygonalComplexes/constructors.gi [hunk 2]
```diff
@@ -941,8 +1027,650 @@ end);
...
+                if j <> false then
+                  Add( oncyc, [r,j] );
+                  # Find the existing neighbours of F
+                  Append(neighs, findNeighs(ud[r],ucode[r][1],F,ucode[r][2]));
+                fi;
+            od;
+            neighs := Set(neighs);
+
+            if Length(oncyc) >= 3 then
+#                Print("face ", F, " is already in 3 cycles\n");
```
Comment (MeikeWeiss) [inline gap/PolygonalComplexes/constructors.gi:1473]: This line can be removed.

**Review by MeikeWeiss — CHANGES_REQUESTED** [review]
Looks great and thanks for all the comments! Here are some few points I have noticed:
1. In my opinion it would be better if there is just once the description of an umbrella descriptor at the begin of the section 4.4 and after that just a reference. 
2. The example for NormedUmbrellaDescriptor is not consistent because, for example, vertex 1 is in the normed umbrella descriptor not an inner vertex, but in the image it is.
3. If I am right __SIMPLICIAL_UmbrellaAdjacentFaces is equal to the local method adjacentFaces in SimplicialSurfaceByUmbrellaDescriptor. It would be nice to use there also __SIMPLICIAL_UmbrellaAdjacentFaces.

**Comment by MeikeWeiss** [comment]
I've noticed that there is also a small mistake in the first picture. Vertices 5 and 10, 6 and 9  and 7 and 8 must be swapped.

**Comment by aniemeyer** [comment]
I have fixed up the last example for NormedUmbrella

**Review by MeikeWeiss — DISMISSED** [review]
Thanks, looks good now!

File: gap/PolygonalComplexes/constructors.gd [hunk 3]
```diff
@@ -1005,3 +1024,183 @@ DeclareOperation( "UmbrellaDescriptorOfSurface", [IsSimplicialSurface] );
...
 #! @EndGroup
+#!
+#!
+#! @BeginGroup
+#! @Description
+#!  A normed umbrella descriptor is a special umbrella descriptor of a
+#!  surface, where an umbrella descriptor is described in 
+#! (<Ref Sect="Section_Constructors_Umbrellas"/>).
+#!  The normed umbrella descriptor of a given umbrella descriptor <A>umdesc</A>
+#!  for the face <A>face</A> and an optional list of neighbours <A>neigh</A>
```
Comment (TomGoertzen) [inline gap/PolygonalComplexes/constructors.gd:1035]: should be: neighbours <A>neighbours</A>  or change parameter to neigh

File: gap/PolygonalComplexes/constructors.gd [hunk 4]
```diff
@@ -1005,3 +1024,183 @@ DeclareOperation( "UmbrellaDescriptorOfSurface", [IsSimplicialSurface] );
...
+#!   (4,9,17,18,11),   (5,6,14,19,12), [ 8, 7, 13 ], [ 8, 9, 17 ], (10,11,18,20,16),
+#!   [ 13, 12, 19 ],    [ 15, 14, 19 ], [ 15, 16, 20 ], [ 17, 18, 20 ] ]
+#! gap>  surf2 := SimplicialSurfaceByUmbrellaDescriptor(ud2);;
+#! gap> IsIsomorphic(surf1,surf2);
+#! true
+#! @EndExampleSession
+#! @Returns a list 
+#! @Arguments umbrelladescriptor, face 
+DeclareOperation( "NormedUmbrellaDescriptor",[IsList,IsPosInt]);
+#! @Arguments umbrelladescriptor, face, neighbours
```
Comment (TomGoertzen) [inline gap/PolygonalComplexes/constructors.gd:1129]: see above

**Review by TomGoertzen — CHANGES_REQUESTED** [review]

File: gap/PolygonalComplexes/constructors.gd (hunk 3 above)
Comment (aniemeyer) [reply gap/PolygonalComplexes/constructors.gd]: Thank you, it is fixed.

**Review by TomGoertzen — APPROVED** [review]

**Review by MeikeWeiss — APPROVED** [review]

**Pull request merged**